tests/unittests/Foundation/NSNotificationQueueTests.mm

```diff
@@ -163,31 +156,62 @@ - (void)scheduleAndAwaitBlock:(void (^)())block {
         queue.attach([[NSNotificationQueue alloc] initWithNotificationCenter:notificationCenter]);
     }];
 
-    __block NSCondition* condition = [[NSCondition new] autorelease];
+    __block _NSBooleanCondition* condition = [[_NSBooleanCondition new] autorelease];
     __block int counter = 0;
     [notificationCenter addObserverForName:s_TestNotificationName
                                     object:nil
                                      queue:nil
                                 usingBlock:^(NSNotification* note) {
                                     ASSERT_OBJCEQ(s_TestNotificationName, note.name);
-                                    [condition lock];
                                     ++counter;
                                     [condition broadcast];
-                                    [condition unlock];
                                 }];
 
-    [condition lock];
-
     [helper scheduleAndAwaitBlock:^{
         [queue enqueueNotification:[NSNotification notificationWithName:s_TestNotificationName object:nil] postingStyle:NSPostWhenIdle];
     }];
 
     [condition waitUntilDate:[NSDate dateWithTimeIntervalSinceNow:20]];
-    [condition unlock];
 
     ASSERT_EQ(1, counter);
 }
 
+TEST(NSNotificationQueue, AllThreePostingStyles) {
+    static NSString* s_TestNotificationName = @(GetTestFullName().c_str());
+    NSNotificationCenter* notificationCenter = [[NSNotificationCenter new] autorelease];
+    NSNotificationQueueTestHelper* helper = [[NSNotificationQueueTestHelper new] autorelease];
+
+    __block StrongId<NSNotificationQueue> queue;
+    [helper scheduleAndAwaitBlock:^{
+        queue.attach([[NSNotificationQueue alloc] initWithNotificationCenter:notificationCenter]);
+    }];
+
+    dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+    std::atomic<int> counter{ 0 };
+    std::atomic<int>* counterp = &counter; // __block couldn't defeat the copy construction of std::atomic?
```

Comment:
yup this is why i had the extremely bad __block std::shared_ptr<std::atomic>>s in my pr yesterday ;;;;